mobile-app/integration_test/news_test.dart

```diff
@@ -0,0 +1,78 @@
+import 'package:flutter/material.dart';
+import 'package:flutter_test/flutter_test.dart';
+import 'package:freecodecamp/main.dart' as app;
+import 'package:freecodecamp/ui/views/news/news-article-post/news_article_post_header.dart';
+import 'package:freecodecamp/ui/views/news/news-feed/news_feed_lazyloading.dart';
+import 'package:integration_test/integration_test.dart';
+import 'package:path/path.dart' as path;
+import 'package:sqflite/sqflite.dart';
+
+void main() {
+  IntegrationTestWidgetsFlutterBinding.ensureInitialized();
+
+  group('News Component', () {
+    testWidgets('should bookmark article', (WidgetTester tester) async {
+      // Start app
+      app.main();
+      await tester.pumpAndSettle();
+
+      // Tap on the first article
+      final Finder firstArticle = find.byType(NewsFeedLazyLoading).first;
+      final ValueKey firstArticleKey =
+          tester.firstWidget<NewsFeedLazyLoading>(firstArticle).key! as ValueKey;
+      expect(firstArticle, findsOneWidget);
+      await tester.tap(firstArticle);
+      await tester.pumpAndSettle();
+
+      // Tap on the bookmark button and store article title and author
+      final Finder bookmarkButton = find.byKey(const Key('bookmark_btn'));
+      final Finder articleTitle = find.byKey(const Key('title'));
+      final Finder articleAuthor = find.descendant(
+          of: find.byType(NewsArticlePostHeader),
+          matching: find.textContaining(RegExp(r'^Written by')));
+      expect(bookmarkButton, findsOneWidget);
+      expect(articleTitle, findsOneWidget);
+      expect(articleAuthor, findsOneWidget);
+      Text title = tester.firstWidget(articleTitle);
+      String author = tester.firstWidget<Text>(articleAuthor).data ?? '';
```

Comment:
Here, for example.

Reply:
The type of `tester.firstWidget<Text>(articleAuthor).data` is `String?` which means that it will return a value of type String or null depending on whether it finds a widget. Also `firstWidget` returns a type of `Widget` or a child of `Widget`(here it's `Text`). `Text.data` type is `String?` by default(https://api.flutter.dev/flutter/widgets/Text/data.html).

Now in terms of using test data or not, there won't be any difference in the type even if the data is present or not. One way of overcoming this nullable type is by typecasting it to non-nullable by adding the postfix operator `!`(an example is in line 22 of same file). Yes, I should have not added a conditional on this line as then when comparing it below there is a chance both values will be null.

More details about darts null safety - https://dart.dev/null-safety

Reply:
In TypeScript terms, this would be `string | null`, right?

Regardless, if you seed the data, then you know what the value is  (even if Dart doesn't).  Then a `null` value is either expected, so should be asserted for, or unexpected and so indicates a bug that should fail the test.

In other words, I'd expect the pseudo-code to be something like this

```
get widget
if (widget.data is null) throw an error 
else do stuff with the widget.data
```

Basically, you almost never want a conditional in a test, because in the event something unexpected happens the test should just fail.